File: backend(Spring Boot)/src/main/java/com/peergrid/backend/controller/PaymentsController.java
package com.peergrid.backend.controller;

import com.peergrid.backend.dto.PaymentRequest;
import com.peergrid.backend.dto.PaymentVerificationRequest;
import com.peergrid.backend.entity.User;
import com.peergrid.backend.repository.UserRepository;
import com.razorpay.Order;
import com.razorpay.RazorpayClient;
import com.razorpay.RazorpayException;
import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.math.BigDecimal;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api/payments")
@CrossOrigin(origins = "*")
public class PaymentsController {

    @Autowired
    private UserRepository userRepository;

    @Value("${razorpay.key.id}")
    private String razorpayKeyId;

    @Value("${razorpay.key.secret}")
    private String razorpayKeySecret;

    private User getAuthenticatedUser() {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        String email = auth.getName();
        return userRepository.findByEmail(email).orElseThrow(() -> new RuntimeException("User not found"));
    }

    @PostMapping("/create-order")
    public ResponseEntity<?> createOrder(@RequestBody PaymentRequest request) {
        try {
            if (razorpayKeyId == null || razorpayKeyId.contains("YOUR_KEY")) {
                return ResponseEntity.badRequest().body("Razorpay keys are not configured in backend.");
            }

            RazorpayClient client = new RazorpayClient(razorpayKeyId, razorpayKeySecret);

            JSONObject options = new JSONObject();
            options.put("amount", request.getAmount().multiply(new BigDecimal(100)).intValue()); // Amount in paise
            options.put("currency", "INR");
            options.put("receipt", UUID.randomUUID().toString());
            options.put("payment_capture", 1); // Auto capture

            Order order = client.orders.create(options);
            String orderId = order.get("id");

            Map<String, Object> response = new HashMap<>();
            response.put("orderId", orderId);
            response.put("keyId", razorpayKeyId);
            return ResponseEntity.ok(response);
        } catch (RazorpayException e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Error creating order: " + e.getMessage());
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Error: " + e.getMessage());
        }
    }

    @PostMapping("/verify-payment")
    public ResponseEntity<?> verifyPayment(@RequestBody PaymentVerificationRequest request) {
        try {
            // Verify Signature
            String payload = request.getOrderId() + "|" + request.getPaymentId();
            String generatedSignature;

            Mac sha256_HMAC = Mac.getInstance("HmacSHA256");
            SecretKeySpec secret_key = new SecretKeySpec(razorpayKeySecret.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            sha256_HMAC.init(secret_key);
            byte[] hash = sha256_HMAC.doFinal(payload.getBytes(StandardCharsets.UTF_8));

            StringBuilder result = new StringBuilder();
            for (byte b : hash) {
                result.append(String.format("%02x", b));
            }
            generatedSignature = result.toString();

            if (generatedSignature.equals(request.getSignature())) {
                User user = getAuthenticatedUser();
                
                // Add Grid Points (1 INR = 10 GP)
                BigDecimal pointsToAdd = request.getAmount().multiply(new BigDecimal(10));
                user.setGridPoints(user.getGridPoints().add(pointsToAdd));
                userRepository.save(user);

                Map<String, Object> response = new HashMap<>();
                response.put("message", "Payment successful");
                response.put("newBalance", user.getGridPoints());
                response.put("pointsAdded", pointsToAdd);
                
                return ResponseEntity.ok(response);
            } else {
                return ResponseEntity.badRequest().body("Invalid payment signature");
            }
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Error verifying payment: " + e.getMessage());
        }
    }
}

File: backend(.NET)/Controllers/PaymentsController.cs
using Microsoft.AspNetCore.Mvc;
using Razorpay.Api;
using PeerGrid.Backend.Data;
using PeerGrid.Backend.Models;
using Microsoft.EntityFrameworkCore;
using System.Security.Claims;
using Microsoft.AspNetCore.Authorization;
using System.Text;

namespace PeerGrid.Backend.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class PaymentsController : ControllerBase
    {
        private readonly IConfiguration _configuration;
        private readonly ApplicationDbContext _context;

        public PaymentsController(IConfiguration configuration, ApplicationDbContext context)
        {
            _configuration = configuration;
            _context = context;
        }

        [HttpPost("create-order")]
        [Authorize]
        public IActionResult CreateOrder([FromBody] PaymentRequest request)
        {
            try
            {
                string key = _configuration["Razorpay:KeyId"];
                string secret = _configuration["Razorpay:KeySecret"];

                if (string.IsNullOrEmpty(key) || key.Contains("YOUR_KEY"))
                {
                    return BadRequest("Razorpay keys are not configured in backend.");
                }

                RazorpayClient client = new RazorpayClient(key, secret);

                Dictionary<string, object> options = new Dictionary<string, object>();
                options.Add("amount", request.Amount * 100);
                options.Add("receipt", Guid.NewGuid().ToString());
                options.Add("currency", "INR");
                options.Add("payment_capture", 1);

                Order order = client.Order.Create(options);
                string orderId = order["id"].ToString();

                return Ok(new { orderId = orderId, keyId = key });
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error creating order: {ex.Message}");
            }
        }

        [HttpPost("verify-payment")]
        [Authorize]
        public async Task<IActionResult> VerifyPayment([FromBody] PaymentVerificationRequest request)
        {
            try
            {
                string key = _configuration["Razorpay:KeyId"];
                string secret = _configuration["Razorpay:KeySecret"];

                // Verify Signature
                string payload = $"{request.OrderId}|{request.PaymentId}";
                string generatedSignature;
                using (var hmac = new System.Security.Cryptography.HMACSHA256(Encoding.UTF8.GetBytes(secret)))
                {
                    var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(payload));
                    generatedSignature = BitConverter.ToString(hash).Replace("-", "").ToLower();
                }

                if (generatedSignature == request.Signature)
                {

                }
                else
                {
                    return BadRequest("Invalid payment signature");
                }


                
                var userIdStr = User.FindFirst(ClaimTypes.NameIdentifier)?.Value;
                if (string.IsNullOrEmpty(userIdStr))
                {
                    return Unauthorized("User ID not found in token");
                }
                var userId = int.Parse(userIdStr);
                var user = await _context.Users.FindAsync(userId);

                if (user != null)
                {
                    // Add Grid Points
                    decimal pointsToAdd = request.Amount * 10; 
                    user.GridPoints += pointsToAdd; 
                    await _context.SaveChangesAsync();

                    // Log the transaction
                    var log = new Log
                    {
                        Type = "Finance",
                        Action = "Payment Success",
                        User = user.Email,
                        Details = $"Payment ID: {request.PaymentId}, Amount: {request.Amount}, GP Added: {pointsToAdd}",
                        Timestamp = DateTime.UtcNow
                    };
                    _context.Logs.Add(log);
                    await _context.SaveChangesAsync();

                    return Ok(new { message = "Payment successful", newBalance = user.GridPoints, pointsAdded = pointsToAdd });
                }

                return BadRequest("User not found");
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error verifying payment: {ex.Message}");
            }
        }
    }

    public class PaymentRequest
    {
        public decimal Amount { get; set; }
    }

    public class PaymentVerificationRequest
    {
        public string OrderId { get; set; }
        public string PaymentId { get; set; }
        public string Signature { get; set; }
        public decimal Amount { get; set; }
    }
}
